### PR TITLE
BUG: Fix memory leak in PY3K buffer code.

### PR DESCRIPTION
**Pull request by vanossj**
I think that `tmp` could leak on python 3 if `_append_char` causes an early exit.

### CODE REVIEW

**Comment by mattip** [comment]
I see your point, but would prefer to localize the use of `tmp` to the block that uses it.

File: numpy/core/src/multiarray/buffer.c [hunk 1]
```diff
@@ -299,29 +299,38 @@ _buffer_format_string(PyArray_Descr *descr, _tmp_string_t *str,
...
 #else
             tmp = name;
+            Py_INCREF(tmp);
 #endif
             if (tmp == NULL || PyBytes_AsStringAndSize(tmp, &p, &len) < 0) {
                 PyErr_Clear();
                 PyErr_SetString(PyExc_ValueError, "invalid field name");
+                Py_XDECREF(tmp);
+                return -1;
+            }
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/buffer.c]: Can't you just `Py_DECREF(tmp)` after this line, and be done with it?

Comment (eric-wieser) [reply numpy/core/src/multiarray/buffer.c]: Ah, nevermind - it has to stick around

File: numpy/core/src/multiarray/buffer.c [hunk 2]
```diff
@@ -299,29 +299,38 @@ _buffer_format_string(PyArray_Descr *descr, _tmp_string_t *str,
...
+            if (_append_char(str, ':') < 0) {
+                Py_DECREF(tmp);   
                 return -1;
             }
-            if (_append_char(str, ':') < 0) return -1;
             while (len > 0) {
                 if (*p == ':') {
-                    Py_DECREF(tmp);
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/buffer.c]: Why did this one move?

**Review by eric-wieser — APPROVED** [review]
A nice follow-up might be to extra an `_append_field_name(_tmp_string_t str, PyObject *name)` helper function which appends `:name:` to the string.

Or we should just convert this all to python, because it would be weird to be bottlenecked by PEP3118 format generation

**Comment by vanossj** [comment]
refactoring the into `_append_field_name` is a bit cleaner

**Comment by eric-wieser** [comment]
I meant that as a separate PR - it's sometimes nice to keep bugfixes and cleanup separate

**Comment by eric-wieser** [comment]
I've pushed an update that squashes your commits into a separate bugfix and refactor. I think this is good to go once tests pass.

**Comment by charris** [comment]
Thanks @vanossj .

**Pull request merged**